File: Week1/7_Cousins_In_Binary_Tree/Solution.cpp
/*
 * Created on Thu May 07 2020
 *
 * Title: Leetcode - Cousins in Binary Tree
 *
 * Author: Vatsal Mistry
 * Web: mistryvatsal.github.io
 */

#include <iostream>
#include <unordered_map>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
};

class Solution {
    public:
    unordered_map<int, pair<int, int>> store;
    void dfs(TreeNode *root, int level, TreeNode *parent) {
        if(root == NULL) {
            return;
        }
        level += 1;
        int parent_val = parent == NULL ? 0 : parent->val;
        pair<int, int> curr = {level, parent_val};
        store[root->val] = curr;
        if(root->left != NULL) {
            dfs(root->left, level, root);
        }
        if(root->right != NULL) {
            dfs(root->right, level, root);
        }
    }

    bool isCousins(TreeNode *root, int x, int y) {
        dfs(root, -1, NULL);
        auto c1 = store[x];
        auto c2 = store[y];
        
        if(c1.first == c2.first and c1.second != c2.second){
            return true;
        }
        return false;
    }
};

File: Week1/7_Cousins_In_Binary_Tree/Solution.py
#
# Created on Fri May 08 2020
#
# Title: Leetcode - Cousins in Binary Tree 
#
# Author: Vatsal Mistry
# Web: mistryvatsal.github.io
#

class TreeNode:
    def __init__(self, val, left, right):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    store = dict()
    def dfs(self, root, level, parent):
        if root is None:
            return False
        level += 1
        parent_val = 0 if parent is None else parent.val
        store[root.val] = [level, parent_val]
        if root.left is not None:
            self.dfs(root.left, level, root)
        if root.right is not None:
            self.dfs(root.right, level, root)

    def isCousin(self, root, x, y):
        dfs(root, -1, None)
        
        if(self.store[x][0] == self.store[y][0] and self.store[x][1] != self.store[y][1]):
            return True
        else:
            return False

